File: frontend/src/services/collaborationService.js
import api from './api';

/**
 * Service to handle community collaboration (comments, follows, bookmarks, shares).
 */

export const addComment = (incidentId, content, parentId = null) => {
  return api.post('/api/comments', { incidentId, content, parentId });
};

export const getComments = (incidentId) => {
  return api.get(`/api/comments/${incidentId}`);
};

export const likeComment = (commentId) => {
  return api.post(`/api/comments/${commentId}/like`);
};

export const followUser = (targetEmail) => {
  return api.post(`/api/follow?targetEmail=${targetEmail}`);
};

export const unfollowUser = (targetEmail) => {
  return api.delete(`/api/follow?targetEmail=${targetEmail}`);
};

export const getFollowing = () => {
  return api.get('/api/follow/following');
};

export const saveReport = (incidentId) => {
  return api.post(`/api/save/${incidentId}`);
};

export const unsaveReport = (incidentId) => {
  return api.delete(`/api/save/${incidentId}`);
};

export const getSavedReportIds = () => {
  return api.get('/api/save/saved');
};

export const logShare = (incidentId) => {
  return api.post(`/api/share/${incidentId}`);
};

File: backend/src/main/java/com/service/CollaborationService.java
package com.service;

import com.model.Comment;
import java.util.List;

/**
 * Service defining collaborative community features including comments, saves, follows, and shares.
 */
public interface CollaborationService {

    Comment addComment(String incidentId, String content, String parentId, String email);

    List<Comment> getComments(String incidentId);

    void likeComment(String commentId, String email);

    void followUser(String targetEmail, String email);

    void unfollowUser(String targetEmail, String email);

    List<String> getFollowing(String email);

    List<String> getFollowers(String email);

    void saveReport(String incidentId, String email);

    void unsaveReport(String incidentId, String email);

    List<String> getSavedReportIds(String email);

    void incrementShareCount(String incidentId);
}
